### PR TITLE
Speed up initial in-memory Soroban state population

### PR DESCRIPTION
**Pull request by drebelsky**
Related to #4902. Note that since that time, state churn has continued, so population now takes ~70s on a dev watcher. This PR changes the live state calculation from going through the buckets one-by-one using a hash map to a k-way merge among all the buckets. The merge is done using a loser tree, which gives us about half as many comparisons as using a heap. Running on a dev watcher (ledger 62163071) speeds up from ~70s to ~30s.

<details>
<summary>Time for 3 runs on upstream vs patch</summary>

```
26.0.2-3192.9b5bee752.noble~do~not~use~in~prd~perftests 300

2026-05-05T18:09:57.683 GB4ZO [Perf INFO] Populated in-memory Soroban state in 31.077 sec
2026-05-05T18:09:57.683 GB4ZO [Perf INFO] Startup state load took 32.777 sec (full=true)

2026-05-05T18:10:33.150 GA6IH [Perf INFO] Populated in-memory Soroban state in 30.273 sec
2026-05-05T18:10:33.151 GA6IH [Perf INFO] Startup state load took 31.619 sec (full=true)

2026-05-05T18:11:08.694 GDECF [Perf INFO] Populated in-memory Soroban state in 30.091 sec
2026-05-05T18:11:08.698 GDECF [Perf INFO] Startup state load took 31.317 sec (full=true)

26.0.2-3190.8a71e20af.noble~perftests 300

2026-05-05T18:05:38.067 GDDEI [Perf INFO] Populated in-memory Soroban state in 68.910 sec
2026-05-05T18:05:38.067 GDDEI [Perf INFO] Startup state load took 70.809 sec (full=true)

2026-05-05T18:07:14.023 GDYRH [Perf INFO] Populated in-memory Soroban state in 70.991 sec
2026-05-05T18:07:14.023 GDYRH [Perf INFO] Startup state load took 73.228 sec (full=true)

2026-05-05T18:08:35.801 GA4XC [Perf INFO] Populated in-memory Soroban state in 74.710 sec
2026-05-05T18:08:35.801 GA4XC [Perf INFO] Startup state load took 76.958 sec (full=true)
```

</details>

Doing the k-way merge also has nicer memory scaling characteristics than the current approach: the amount of memory we use scales with the live state + number of buckets, instead of the current approach that scales with churn (e.g., on a local run, peak memory was ~4.10 GB vs ~7.47 GB).

Additionally, the PR disables bucket merges until after the in-memory state is populated.

### CODE REVIEW

File: src/bucket/BucketListSnapshot.cpp [hunk 1]
```diff
@@ -708,6 +708,198 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+        mKey = getBucketLedgerKey(mEntry);
+        if (mKey.type() > mType)
+        {
+            break;
+        }
+
+        auto type = mEntry.type();
+        releaseAssert(type == LIVEENTRY || type == INITENTRY ||
+                      type == DEADENTRY);
+        return true;
```
[the remark this answers is not in the document]
Comment (drebelsky) [reply src/bucket/BucketListSnapshot.cpp]: `getRangeForType()` is exact, but I changed this so that it more closely matches the path in `SearchableLiveBucketListSnapshot::scanForEntriesOfType` (checks that the type matches before returning).

File: src/bucket/LedgerCmp.h [hunk 2]
```diff
@@ -123,6 +124,10 @@ struct LedgerEntryIdCmp
     }
 };
 
+// Like LedgerEntryIdCmp, but only compares LedgerKeys, and does a 3-way
+// comparison instead of a less-than.
+std::partial_ordering compareLedgerKeys(LedgerKey const& a, LedgerKey const& b);
```
Comment (bboston7) [inline src/bucket/LedgerCmp.h]: Why a partial ordering? Does a total ordering not exist for ledger keys?

File: src/bucket/BucketListSnapshot.cpp [hunk 3]
```diff
@@ -708,6 +708,198 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+    }
+    return false;
+}
+} // namespace
+
+void
+SearchableLiveBucketListSnapshot::scanForLiveEntriesOfType(
+    LedgerEntryType type,
+    std::function<void(LedgerEntry const&, LedgerKey const&)> callback) const
+{
```
Comment (bboston7) [inline src/bucket/BucketListSnapshot.cpp:765]: This should probably have a `ZoneScoped`

File: src/bucket/BucketListSnapshot.cpp [hunk 4]
```diff
@@ -708,6 +708,198 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+        while (i > 0)
+        {
+            if (leftWins(tree[i], winner))
+            {
+                std::swap(tree[i], winner);
+            }
+            i /= 2;
+        }
+
+        tree[0] = winner;
```
Comment (bboston7) [inline src/bucket/BucketListSnapshot.cpp:905]: Can you please add some comments to this? It's a little hard to figure out what's going on here. This is the actual k-way merge, right?

**Review by bboston7 — COMMENTED** [review]
I'm by no means an expert on the bucket subsystem, but the algorithm implementation looks correct to me. I just had a few questions along the way.

File: src/bucket/LedgerCmp.h (hunk 2 above)
Comment (drebelsky) [reply src/bucket/LedgerCmp.h]: The compare delegates to the `operator <=>` from `xdrpp` (which is important since we do need to match how `LedgerEntryIdCmp` does the ordering (and since, e.g., the value type in `ScVal` for `CONTRACT_DATA` is nested pretty deeply). I'll open a PR in `xdrpp` to fix it.

File: src/bucket/BucketListSnapshot.cpp [hunk 5]
```diff
@@ -708,6 +708,203 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+        {
+            return true;
+        }
+    }
+    return false;
+}
+} // namespace
+
+void
+SearchableLiveBucketListSnapshot::scanForLiveEntriesOfType(
```
Comment (SirTyson) [inline src/bucket/BucketListSnapshot.cpp:762]: I think this looks generally correct, but it's adding a good amount of complexity, I'd like to see a unit test specifically for this scanning function. Before it was pretty straight forward and indirectly tested, but given the k-way merge I think a more explicit test is warranted. Maybe we can test some of the loser tree edge cases, like a degenerate merge with just 1 bucket, 2 buckets, and some non powers of two. It might also be a good idea to hook this into the randomized bucket testing infra LedgerStateSnapshotTests,cpp or BucketIndexTests.cpp, where we just make sure we hit all the entries properly.

File: src/bucket/BucketListSnapshot.cpp [hunk 6]
```diff
@@ -708,6 +708,203 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+
+        if (!iter.advance())
+        {
+            tree[index + numIterators] = exhausted;
+        }
+        int winner = tree[index + numIterators];
+
+        // Update tournament up the tree to the root
+        int i = (index + numIterators) / 2;
+        while (i > 0)
```
Comment (SirTyson) [inline src/bucket/BucketListSnapshot.cpp]: Nit: This while could be a for loop, which to me reads a little cleaner.

File: src/bucket/BucketListSnapshot.cpp (hunk 4 above)
Comment (SirTyson) [reply src/bucket/BucketListSnapshot.cpp:905]: Agreed, it's still a little opaque.

File: src/bucket/BucketListSnapshot.cpp [hunk 7]
```diff
@@ -708,6 +708,203 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+        {
+            tree[numIterators + i] = i;
+        }
+        else
+        {
+            tree[numIterators + i] = exhausted;
+        }
+    }
+
+    auto leftWins = [&iterators](int leftIndex, int rightIndex) -> bool {
```
Comment (SirTyson) [inline src/bucket/BucketListSnapshot.cpp:831]: More comments here would be helpful, this indicates that the smaller, newer version index wins, right?

File: src/ledger/InMemorySorobanState.cpp [hunk 8]
```diff
@@ -456,75 +456,26 @@ InMemorySorobanState::initializeStateFromSnapshot(
...
-            }
-
-            releaseAssertOrThrow(be.type() == LIVEENTRY ||
-                                 be.type() == INITENTRY);
-            auto lk = LedgerEntryKey(be.liveEntry());
-            releaseAssertOrThrow(lk.type() == expectedType);
-            return deletedKeys.find(lk) == deletedKeys.end();
+        auto contractDataHandler = [this](LedgerEntry const& le,
+                                          LedgerKey const&) {
+            createContractDataEntry(le);
```
Comment (SirTyson) [inline src/ledger/InMemorySorobanState.cpp:461]: Something I noticed in `createContractDataEntry`, we call xdr_size on le, which is a full recursive traversal of the SCVal for data types. Can we return this from `readOne` and just pipe it through? I remember you mentioned that XDR decode was a pretty significant bottleneck, that might be an easy win.

File: src/ledger/InMemorySorobanState.cpp [hunk 9]
```diff
@@ -456,75 +456,26 @@ InMemorySorobanState::initializeStateFromSnapshot(
...
-                deletedKeys.insert(be.deadEntry());
-                return false;
-            }
-
-            releaseAssertOrThrow(be.type() == LIVEENTRY ||
-                                 be.type() == INITENTRY);
-            auto lk = LedgerEntryKey(be.liveEntry());
-            releaseAssertOrThrow(lk.type() == expectedType);
-            return deletedKeys.find(lk) == deletedKeys.end();
+        auto contractDataHandler = [this](LedgerEntry const& le,
```
Comment (SirTyson) [inline src/ledger/InMemorySorobanState.cpp:459]: Is it worth reserving `mContractDataEntries` and `mCotnratCodeEntries` here based on `getRangeForType` for the buckets?

File: src/bucket/BucketListSnapshot.cpp [hunk 10]
```diff
@@ -708,6 +708,203 @@ SearchableLiveBucketListSnapshot::scanForEntriesOfType(
...
+    getEntry() const
+    {
+        return mEntry;
+    }
+    LedgerKey const&
+    getKey() const
+    {
+        return mKey;
+    }
+    bool advance();
```
Comment (SirTyson) [inline src/bucket/BucketListSnapshot.cpp]: Nit: Stylistically this is a little weird. Can we either inline the advance or just move the declaration to .h?

**Review by SirTyson — CHANGES_REQUESTED** [review]
Thanks for working on this, it looks like a nice change! It does look like we've added a fair bit of complexity to the scanning function, so I'd like to see some unit tests on the new algo and some randomized testing to make sure we've covered all our bases.

**Comment by drebelsky** [comment]
Build failure was from being on an out-of-date branch, rebasing

File: src/ledger/InMemorySorobanState.cpp (hunk 8 above)
Comment (drebelsky) [reply src/ledger/InMemorySorobanState.cpp:461]: This ends up not being that expensive, I think it's probably not worth the added complexity. Running locally on ledger 63389567 with the comparison having `xdr::xdr_size(...)` replaced with `100` in `contractCodeSizeForRent` and `updateStateSizeOnEntryUpdate` gives these results.

```
Current

2026-07-14T16:14:07.328 GBYAB [Perf INFO] Populated in-memory Soroban state in 22.051 sec
2026-07-14T16:14:07.328 GBYAB [Perf INFO] Startup state load took 22.820 sec (full=true)

2026-07-14T16:14:32.035 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.577 sec
2026-07-14T16:14:32.036 GBYAB [Perf INFO] Startup state load took 22.302 sec (full=true)

2026-07-14T16:14:56.716 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.793 sec
2026-07-14T16:14:56.716 GBYAB [Perf INFO] Startup state load took 22.507 sec (full=true)

w/o xdr::xdr_size

2026-07-14T16:18:42.282 GBYAB [Perf INFO] Populated in-memory Soroban state in 20.934 sec
2026-07-14T16:18:42.283 GBYAB [Perf INFO] Startup state load took 21.667 sec (full=true)

2026-07-14T16:19:06.451 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.362 sec
2026-07-14T16:19:06.451 GBYAB [Perf INFO] Startup state load took 22.069 sec (full=true)

2026-07-14T16:19:30.936 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.556 sec
2026-07-14T16:19:30.936 GBYAB [Perf INFO] Startup state load took 22.293 sec (full=true)
```

File: src/ledger/InMemorySorobanState.cpp (hunk 9 above)
Comment (drebelsky) [reply src/ledger/InMemorySorobanState.cpp:459]: I'm a little reticent to do it based on `getRangeForType` since this just gives us byte offsets and not number of entries (and since the extra storage we use will be persistent after the load). Nico did mention one idea of storing the current size in storestate, which would at least let reloads be faster.

Comment (drebelsky) [reply src/ledger/InMemorySorobanState.cpp:459]: Update: I ran locally with exact sizing and the speed-up is negligible

<details>
<summary>diff</summary>

```diff
diff --git a/src/ledger/InMemorySorobanState.cpp b/src/ledger/InMemorySorobanState.cpp
index cf0e1e690..24549d16f 100644
--- a/src/ledger/InMemorySorobanState.cpp
+++ b/src/ledger/InMemorySorobanState.cpp
@@ -455,6 +455,9 @@ InMemorySorobanState::initializeStateFromSnapshot(
     auto ledgerVersion = lclHeader.ledgerVersion;
     if (protocolVersionStartsFrom(ledgerVersion, SOROBAN_PROTOCOL_VERSION))
     {
+        mContractDataEntries.reserve(3362751);
+        mContractCodeEntries.reserve(1937);
+
         auto sorobanConfig = SorobanNetworkConfig::loadFromLedger(applyView);
         auto contractDataHandler = [this](LedgerEntry const& le,
                                           LedgerKey const&) {
@@ -476,6 +479,9 @@ InMemorySorobanState::initializeStateFromSnapshot(
         applyView.scanCurrentLiveEntriesOfType(TTL, ttlHandler);
         applyView.scanCurrentLiveEntriesOfType(CONTRACT_CODE,
                                                contractCodeHandler);
+
+        releaseAssert(mContractDataEntries.size() == 3362751);
+        releaseAssert(mContractCodeEntries.size() == 1937);
     }
 
     mLastClosedLedgerSeq = lclHeader.ledgerSeq;
```
</details>

On ledger 63389567

```
w/o change

2026-07-14T16:14:07.328 GBYAB [Perf INFO] Populated in-memory Soroban state in 22.051 sec
2026-07-14T16:14:07.328 GBYAB [Perf INFO] Startup state load took 22.820 sec (full=true)

2026-07-14T16:14:32.035 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.577 sec
2026-07-14T16:14:32.036 GBYAB [Perf INFO] Startup state load took 22.302 sec (full=true)

2026-07-14T16:14:56.716 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.793 sec
2026-07-14T16:14:56.716 GBYAB [Perf INFO] Startup state load took 22.507 sec (full=true)

w/ change

2026-07-14T16:40:19.939 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.125 sec
2026-07-14T16:40:19.939 GBYAB [Perf INFO] Startup state load took 21.884 sec (full=true)

2026-07-14T16:40:43.969 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.151 sec
2026-07-14T16:40:43.969 GBYAB [Perf INFO] Startup state load took 21.888 sec (full=true)

2026-07-14T16:41:07.882 GBYAB [Perf INFO] Populated in-memory Soroban state in 21.039 sec
2026-07-14T16:41:07.882 GBYAB [Perf INFO] Startup state load took 21.757 sec (full=true)
```

File: src/bucket/test/BucketIndexTests.cpp [hunk 11]
```diff
@@ -46,6 +46,20 @@ class BucketIndexTest
...
+    // tracks every generated batch. Not maintained by
+    // insertSimilarContractDataKeys, and only insertions are tracked for
+    // BucketIndexPoolShareTest (its update/delete path bypasses this map).
+    UnorderedMap<LedgerKey, LedgerEntry> mAllEntries;
+
+    // Number of ledgers after creation at which generated TTLs expire. Must
+    // exceed the number of ledgers buildBucketList closes
+    // (~levelHalf(mLevelsToBuild - 1)), or eviction would start deleting
+    // entries behind the harness maps' backs.
+    static constexpr uint32_t TTL_DURATION = 10'000;
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp]: Can't we just make this arbitrarily huge instead of something reachable like 10k and remove the asserts?

File: src/bucket/test/BucketIndexTests.cpp [hunk 12]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+                    REQUIRE(key == LedgerEntryKey(entry));
+
+                    // Each key should only be emitted once
+                    REQUIRE(found.emplace(key, entry).second);
+                });
+            REQUIRE(found == expected);
+        }
+    };
+
+    SECTION("soroban only")
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp]: Nit: I don't think this has to be a section, just a {} block makes the test faster.

File: src/bucket/test/BucketIndexTests.cpp [hunk 13]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+        auto f = [&](Config& cfg) {
+            BucketIndexTest test{cfg};
+            test.buildMultiVersionTest(/*sorobanOnly=*/true);
+            verifyScans(test, {CONTRACT_DATA, CONTRACT_CODE, TTL});
+
+            // No OFFER entries were generated, so the scan should never invoke
+            // the callback
+            ImmutableLedgerView ledgerView =
+                test.getApp().getLedgerManager().copyImmutableLedgerView();
+            ledgerView.scanCurrentLiveEntriesOfType(
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp:1706]: For completeness here why not call all non-soroban types?

File: src/bucket/test/BucketIndexTests.cpp [hunk 14]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+            view.scanCurrentLiveEntriesOfType(
+                type,
+                [](LedgerEntry const&, LedgerKey const&) { REQUIRE(false); });
+        };
+
+        SECTION("k disjoint buckets")
+        {
+            for (size_t k = 0; k < LiveBucketList::kNumLevels; ++k)
+            {
+                INFO("k = " << k);
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp:1818]: INFO seems noisy for CI

File: src/bucket/test/BucketIndexTests.cpp [hunk 15]
```diff
@@ -46,6 +46,20 @@ class BucketIndexTest
     UnorderedMap<LedgerKey, LedgerEntry> mContractCodeEntries;
     UnorderedMap<LedgerKey, LedgerEntry> mContractDataEntries;
 
+    // Complete mapping of every key currently live in the BucketList
+    // (including TTLs), maintained across inserts, updates, and deletes.
+    // Unlike mTestEntries, which only tracks a sampled subset of entries, this
+    // tracks every generated batch. Not maintained by
+    // insertSimilarContractDataKeys, and only insertions are tracked for
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp]: Hmm this field seems kind of weird. The test is already complicated, and I don't think we should add something called `mAllEntries` if it's not maintained in some functions or test cases. Can we either make it such that it's actually correct throughout the test suite, or scope it better to our new tests that rely on it?

File: src/bucket/test/BucketIndexTests.cpp [hunk 16]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+        testAllIndexTypes(f);
+    }
+
+    SECTION("all types")
+    {
+        auto f = [&](Config& cfg) {
+            BucketIndexTest test{cfg};
+            test.buildMultiVersionTest();
+
+            // ACCOUNT is skipped because genesis entries (the root account)
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp]: I think I'd rather bootstrap the `mAllEntries` field with the source account. I don't care about network configs since we don't generate those, but we do generate account load.

File: src/bucket/test/BucketIndexTests.cpp [hunk 17]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+        };
+
+        auto requireNoCallback = [](ImmutableLedgerView const& view,
+                                    LedgerEntryType type) {
+            view.scanCurrentLiveEntriesOfType(
+                type,
+                [](LedgerEntry const&, LedgerKey const&) { REQUIRE(false); });
+        };
+
+        SECTION("k disjoint buckets")
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp:1808]: It's a little hard to follow what bucket list structure you're building in these test cases. Could you maybe add a block comment at the beginning of each section discussing what the structure is and why it's interesting? Then maybe a comment or two as you build up the structure, just so the reader can follow along and check the bucket looks as intended.

File: src/bucket/test/BucketIndexTests.cpp [hunk 18]
```diff
@@ -1587,4 +1620,284 @@ TEST_CASE("getRangeForType bounds verification", "[bucket][bucketindex]")
...
+                  withSeq(inEveryBucket, 90)})
+            {
+                expected.emplace(LedgerEntryKey(e), e);
+            }
+
+            auto view = makeViewFromBuckets(buckets);
+            REQUIRE(scan(view, OFFER) == expected);
+        }
+
+        SECTION("same bucket at two slots")
```
Comment (SirTyson) [inline src/bucket/test/BucketIndexTests.cpp:1944]: This is a good test case! While it works, I noticed we do re-use the same stream object, which seems like a footgun, as intuitively I would expect separate streams when we're comparing separate logical levels of the BucketList. Is there an easy way to harden this?

File: src/bucket/LedgerCmp.cpp [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        return cmp;
+    }
+    return lexComparePartial(std::forward<U>(args)...);
+}
+} // namespace
+
+namespace stellar
+{
+std::partial_ordering
+compareLedgerKeys(LedgerKey const& a, LedgerKey const& b)
```
Comment (SirTyson) [inline src/bucket/LedgerCmp.cpp]: Having two basically identical compare functions with different implementations seems like a footgun. Can we refactor this to use a templated internal implementation, such that the logic is unified to a single place? I'm worried we could forget to update one of the comparators, and it would be difficult to spot the divergence in tests.

**Review by SirTyson — CHANGES_REQUESTED** [review]
I appreciate the added tests! The BucketList is pretty complex, and the test set up is a bit hard to follow, so I'd appreciate some comments. I also think we should unify the comparison functions, as it's too easy to forget to update one of them if we ever add new types. Implementation wise though, looks good! Thanks for investigating my follow up perf ideas, sorry they didn't seem to be worth the squeeze :(

File: src/bucket/test/BucketIndexTests.cpp (hunk 14 above)
Comment (drebelsky) [reply src/bucket/test/BucketIndexTests.cpp:1818]: INFO only gets printed out if the test case fails

File: src/bucket/test/BucketIndexTests.cpp (hunk 18 above)
Comment (SirTyson) [reply src/bucket/test/BucketIndexTests.cpp:1944]: I think this is a legit bug, it looks like this actually does fail if there's a slightly more interesting repeat bucket:

```
        SECTION("same bucket at non-adjacent slots with shadowing")
        {
            /* Put the same tombstone bucket A in both the newest and oldest
             * slots, with a conflicting live bucket B between them. Each
             * logical slot must have an independent stream cursor: the newest
             * A should shadow B for every key, regardless of the older copy of
             * A.
             *
             * A: DEAD(key1), DEAD(key2)
             * B:             LIVE(key2)
             * A: DEAD(key1), DEAD(key2)
             */
            auto offers =
                LedgerTestUtils::generateValidUniqueLedgerEntriesWithTypes(
                    {OFFER}, 2, generatedKeys);
            std::sort(offers.begin(), offers.end(),
                      [](LedgerEntry const& lhs, LedgerEntry const& rhs) {
                          return LedgerEntryIdCmp{}(lhs.data, rhs.data);
                      });

            std::vector<LedgerKey> deadKeys;
            for (auto const& offer : offers)
            {
                deadKeys.emplace_back(LedgerEntryKey(offer));
            }

            auto bucketA = makeBucket({}, {}, deadKeys, /*pad=*/false);
            auto bucketB =
                makeBucket({}, {offers.back()}, {}, /*pad=*/false);

            auto view = makeViewFromBuckets({bucketA, bucketB, bucketA});
            requireNoCallback(view, OFFER);
        }
```

The test case does fail, am I missing something or calling your test helpers incorrectly?

**Review by SirTyson — CHANGES_REQUESTED** [review]
Changes look good! I do think the file stream issue actually is a correctness bug, let me know if my unit test is missing something but it seems like it fails.

File: src/bucket/test/BucketIndexTests.cpp (hunk 18 above)
Comment (drebelsky) [reply src/bucket/test/BucketIndexTests.cpp:1944]: I misunderstood your first comment. Yeah, I agree it is a bit of a footgun. Updated (and added the test).

File: src/bucket/BucketListSnapshot.cpp [hunk 20]
```diff
@@ -791,10 +791,11 @@ SearchableLiveBucketListSnapshot::scanForLiveEntriesOfType(
             return Loop::INCOMPLETE;
         }
 
-        auto& stream = getStream(bucket);
+        XDRInputFileStream stream;
```
Comment (SirTyson) [inline src/bucket/BucketListSnapshot.cpp]: Can we add a comment here about why we need a new stream here and can't use `getStream` like we do everywhere else?

**Review by SirTyson — DISMISSED** [review]
LGTM! Can we add the last commit and sqaush, then should be ready top merge?

**Review by SirTyson — APPROVED** [review]
